File: src/wdownsample.c
/*
 * wdownsample.c - subset a matrix, tile, and compute the mean of each tile 
 *
 * - subset is specified by integer indices, which may contain
 *   replicate selections, or missing values (represented by -1)
 *   and may be arbitrarily re-ordered
 * - averaging is done over integer-sized windows for rows and columns
 * - the last window is padded by missing indices
 * - caller responsible in ensuring irow and icol are within 0:(mx-1)
 *   and 0:(nx-1)
 */

#include <stdio.h>

#include <math.h>

#define NULL_ID -1   // index to nothing

void
wdownsample (
  int *dims, // mx, nx, my, ny, mw, nw,
  int *irow, // row indices (of length my * mw)
  int *icol, // col indices (of length ny * nw)
  double *x, // input matrix of size mx * nx [column vector is contiguous]
  double *wx, // input weights
  double *y,  // output matrix of size my * ny
  double *wy, // output weights
  double *wrow,  // sum of non-NA indices for each tile row
  double *wcol   // sum of non-NA indices for each tile column
  )
{
  int mx = dims[0];  // input size
  int nx = dims[1];
  int my = dims[2];  // output size
  int ny = dims[3];
  int mw = dims[4];  // window size
  int nw = dims[5];

  for(int i = 0; i < my; i++ )
    {
    wrow[i] = 0;
    for(int u = 0; u < mw; u++ )
      if( irow[i*mw + u] != NULL_ID )
        wrow[i]++;
    }

  for(int j = 0; j < ny; j++ )
    {
    wcol[j] = 0;
    for(int v = 0; v < nw; v++ )
      if( icol[j*nw + v] != NULL_ID  )
        wcol[j]++;
    }

  for(int j = 0; j < ny; j++ )
    {
    if( wcol[j] == 0 )
      {
      for(int i = 0; i < my; i++ )
        {
        y[i + j*my] = 0;
        wy[i + j*my] = 0;
        }
      continue;
      }
    for(int i = 0; i < my; i++ )
      {
      if( wrow[i] == 0 )
        {
        y[i + j*my] = 0;
        wy[i + j*my] = 0;
        continue;  
        }

      double sy = 0;
      double sw = 0;

      for(int v = 0; v < nw; v++ )
        {
        int jx = icol[j*nw + v];
        if( jx == NULL_ID ) continue;
        for(int u = 0; u < mw; u++ )
          {
          int ix = irow[i*mw + u];
          if(ix == NULL_ID ) continue;
          double w = wx[ ix + jx * mx ];
          sy += w * x[ ix + jx * mx ];
          sw += w;
          }
        }
      
      wy[i + j*my] = sw;
      y[i + j*my] = sw > 0 ?  sy/sw: 0;
      }
    }
  return;
}

File: R/wdownsample.R
wdownsample <- function(
  x,w=NULL,
  irow=NULL,      # row index (integer or characters)
  icol=NULL,      # col index (integer or characters)
  tilesize=c(1,1) # tile size
  )
{
  if( class(x)[1]=="tile" )
    {
    w <- data$w
    x <- data$x
    }
  else if(is.null(w)) 
    {
    return( downsample(x,irow,icol,tilesize))
    }

  mx <- nrow(x)
  nx <- ncol(x)
  if( is.null(irow) ) irow <- 1:mx
  if( is.null(icol) ) icol <- 1:nx

  if( !is.numeric(tilesize) )
    stop("tilesize has to be numeric")
  tilesize <- as.integer(tilesize)  
  if( any((tilesize < c(1,1)) | (tilesize > c(mx,nx))) )
    stop("tilesize out of range.")

  if( is.character(irow) )
    {
    if( is.null(rownames(x)) )
      stop("`x` has no rownames to match `irow`")
    irow <- match( irow, rownames(x) )
    }
  irow <- ifelse( is.na(irow) | irow > mx | irow < 1, 0, irow )

  # pad with null_id
  irow <- c(irow, rep(0, (tilesize[1] - length(irow)) %% tilesize[1]))

  if( is.character(icol) )
    {
    if( is.null(colnames(x)) )
      stop("`x` has no colnames to match `icol`")
    icol <- match( icol, colnames(x) )
    }
  icol <- ifelse( is.na(icol) | icol > nx | icol < 1, 0, icol )

  # pad with null_id
  icol <- c(icol, rep(0, (tilesize[2] - length(icol)) %% tilesize[2]))
  
  my <- length(irow) / tilesize[1]
  ny <- length(icol) / tilesize[2]

  r <- .C("wdownsample",
    as.integer(c(mx,nx,my,ny,tilesize)),
    as.integer(irow-1),
    as.integer(icol-1),
    as.double(x),
    as.double(w),
    y = double(my * ny),
    wy = double(my * ny),
    wrow = double(my),
    wcol = double(ny),
    NAOK=TRUE,
    PACKAGE="nclust")

  dim(r$y) <- c(my,ny)
  attr(r$y,"wrow") <- r$wrow
  attr(r$y,"wcol") <- r$wcol

  dim(r$wy) <- c(my,ny)
  attr(r$wy,"wrow") <- r$wrow
  attr(r$wy,"wcol") <- r$wcol

  list(y=r$y,wy=r$wy/outer(r$wrow,r$wcol))
}

wrow <- function(x) attr(x,"wrow")

wcol <- function(x) attr(x,"wcol")
